ast_builder.py

```diff
@@ -4,13 +4,32 @@
 class ASTBuilder(PTNodeVisitor):
 
     def visit_robolang(self, node, children):
-        return BlockNode(children)
+        #Since the only time that this is being visited by itself is at the beginning of the program
+        #default count is 1, for "run the program once"
+        return BlockNode(1, children)
 
     def visit_command(self, node, children):
-        return CommandNode(children[0])
+        #R / L / F by themselves
+        if len(children) == 1:
+            return CommandNode(1, children[0])
+        #count + R / count + L / count + F
+        elif len(children) == 2:
+            return CommandNode(children[0], children[1])
+        #block [RF] (with or without count)    
+        else:
+            if(children[0] == "["):
+                #count is default of 1, as it is not specified
+                return BlockNode(1, children[1])
+            else:
+                #count is specified as child[0]
+                return BlockNode(children[0], children[2])
```

Comment:
Something I'd like all students to do is to work on developing a sense of when something is not the way it should be.

When I look at this function, my spidy-sense is in overdrive. Partly it's the duplication, and partly it's the fact that you're doing syntactic analysis, which is what the PEG parser is supposed to do.

If, instead, you'd defined the grammar like this:

~~~ py
command =
      count? motion
  /   count? turn
  /   pick_drop
  /   count? block

block =
  "[" robolang "]"
~~~

Then the AST builder would be

~~~ py
    def visit_command(self, node, children):
        if len(children) == 1:
            return CommandNode(children[0], 1)
        else:
            return CommandNode(children[1], children[0])

    def visit_block(self, node, children):
        return children[0]
~~~

Which is both simpler and reflects the actual grammatical structure of the language.

# -3